pkg/pillar/cmd/zedrouter/zedrouter.go

```diff
@@ -511,14 +524,19 @@ func Run(ps *pubsub.PubSub, loggerArg *logrus.Logger, logArg *base.LogObject) in
 			subAssignableAdapters.ProcessChange(change)
 
 		case change := <-subAppNetworkConfig.MsgChan():
-			// If we have an NetworkInstanceConfig process it first
-			checkAndProcessNetworkInstanceConfig(&zedrouterCtx)
+			// in modify path
+			// If we have NetworkInstanceConfig event, process it first
+			checkAndProcessNetworkInstanceConfig(change,
+				&zedrouterCtx)
 			subAppNetworkConfig.ProcessChange(change)
 
 		case change := <-subDeviceNetworkStatus.MsgChan():
 			subDeviceNetworkStatus.ProcessChange(change)
 
 		case change := <-subNetworkInstanceConfig.MsgChan():
+			// in delete path
+			// If we have AppNetworkConfig event, process it first
+			checkAndProcessAppNetworkConfig(change, &zedrouterCtx)
```

Comment:
I don't see how this can solve the crash/assertion since normally if the controller deletes everything the domainmgr will halt the guest VM which can take minutes, and not until that is done will you see the AppNetworkConfig being deleted.
Hence the delete of the NetworkInstanceConfig can happen minutes before the delete of the AppNetworkConfigs.

I think the delete code in networkinstance.go needs to defer the removal (of the NetworkInstanceStatus and its info) until the Vifs is empty, meaning that all of the AppNetworkConfig whch use the NIS have been at least deactivated.

Reply:
Thanks for removing this